File: Python/一知半解/Hard/23. Merge k Sorted Lists.py
import heapq
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class PriorityListNode(ListNode):
    def __lt__(self, other):
        return self.val < other.val
 
class Solution:
    def mergeKLists(self, lists: [[ListNode]]) -> [ListNode]:
        pq = []
        for listt in lists:
            while(listt):
                pL = PriorityListNode(listt.val)
                heapq.heappush(pq,(pL.val,pL))
                listt = listt.next
        start = heapq.heappop(pq)[1]     
        if (not start):
            return None
        answer = start      
        while(pq):
            start.next = heapq.heappop(pq)[1] 
            start = start.next 
        start.next = None
        return answer
            

File: C++/一知半解/Hard/23. Merge k Sorted Lists.cpp
#include<iostream>
#include <bits/stdc++.h>
#include <vector>


struct ListNode{
    int val;
    ListNode* next;
    ListNode():val(0),next(nullptr){}
    ListNode(int val):val(val),next(nullptr){}
    ListNode(int val,ListNode* next):val(val),next(next){}

};

class Solution {
public:
    ListNode* mergeKLists(std::vector<ListNode*>& lists) {
        std::priority_queue<int,std::vector<int>,std::greater<int>> minHeap ;
        ListNode* dummyNode = new ListNode();
        for(ListNode* listNode : lists){
            while(listNode){
                minHeap.emplace(listNode->val);
                listNode = listNode->next;
            }
        }
        while(!minHeap.empty()){
            int node = minHeap.top();
            dummyNode->next = new ListNode(node);
            dummyNode = dummyNode->next;
            minHeap.pop();
        }
        return dummyNode->next;
        
    }
};